### ISSUE TITLE
Default Rules file is not suited for clean URLs

### ISSUE THREAD

**Issue by ddfreyne** [nanoc/nanoc#588]
The Rules file is not well-suited for clean URLs. A file `content/about.html` will be routed to `output/about.html` rather than `output/about/index.html`.

For the record, the default Rules file in nanoc 4 currently looks like this:

```ruby
#!/usr/bin/env ruby

compile '/**/*.html' do
  filter :erb
  layout '/default.*'
end

compile '/**/*' do
end

route '/**/*' do
  item.identifier.to_s
end

layout '/**/*', :erb
```

**Comment by ddfreyne** [comment]
Here’s an alternative Rules file that is more extensible. Important changes:

* It generates clean URLs for HTML files.

* It also includes a `.rdoc` extension, because I believe it is a good idea to introduce the concept of filtering by extension, and RDoc is part of Ruby anyway, so there are no external dependencies.

* It handles /index.* files properly.

```ruby
compile '/**/*.html' do
  layout '/default.*'
end

compile '/**/*.rdoc' do
  filter :rdoc
  layout '/default.*'
end

route '/**/*.{html,rdoc}' do
  case item.identifier
  when '/index.*'
    '/index.html'
  else
    item.identifier.without_ext + '/index.html'
  end
end

route '/**/*' do
  item.identifier.to_s
end

layout '/**/*', :erb
```

**Comment by bobthecow** [comment]
I'd probably put a rule (commented out) for handling `*.md`, just because that's likely to be the most common question.

**Comment by TheLonelyGhost** [comment]
:+1: for `*.md` rule in default Rules file, though that's likely off-topic for this issue.

Perhaps I'm behind on the history of nanoc, but why was this feature implemented in the first place? Is the appearance of `/about/` in the browser's navigation bar really cleaner than `/about.html`? What about the effect it has on cluttering the filesystem with directories holding exactly one file?

Both options seem clean to me and the latter indicates the format on the actual URL while making no assumptions about the index file on the HTTP server.

**Comment by bobthecow** [comment]
http://www.w3.org/Provider/Style/URI.html

**Comment by ddfreyne** [comment]
I prefer `http://nanoc.ws/about/` over `http://nanoc.ws/about.html`. The former is easier to read, and also easier to remember/type.

With nanoc’s routing rules, it’s still pretty easy to *not* use this approach (some people prefer `about.html`).

**Comment by ddfreyne** [comment]
Potential fix in #591.

**Comment by ddfreyne** [comment]
Fixed by #591.
